extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/shuffle/DurableStorageOutputChannelFactory.java

```diff
@@ -112,31 +132,52 @@ public OutputChannel openNilChannel(int partitionNumber)
       throw new ISE(
           e,
           "Unable to create empty remote output of workerTask[%s] stage[%d] partition[%d]",
-          workerTaskId,
+          workerNumber,
           stageNumber,
           partitionNumber
       );
     }
   }
 
+  public void markAsSuccess(int partitionNumber) throws IOException
+  {
+    String oldPath = getPartitionOutputFileNameForTask(controllerTaskId, workerNumber, stageNumber, partitionNumber, taskId);
+    storageConnector.moveFile(oldPath, StringUtils.format("%s%s", oldPath, SUCCESSFUL_SUFFIX));
```

Comment:
This is a very expensive call in s3. IIRC you basically copy from old path to new path. Its not just a metadata update as in the case of HDFS. 

IMHO we might need a better approach here. 

Just create a __SUCCESS file in the dir ?

Reply:
That way we can also use the timestamp of that file for tiebreaker.

Reply:
Any alternatives for the S3 client? The current implementation has copy and delete in S3. I didn't find a rename operation.

Reply:
I donot think there is support from AWS regarding this. So its more of a backend thing rather than a client thing